[GSoC] Added ```Q.gt(x, y)``` ⇒ ```Q.ge(x, y)``` and ```Q.lt(x, y)``` ⇒ ```Q.le(x, y)``` implications in assumptions

krishnavbajoria02 opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->


<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234" (see
https://tinyurl.com/auto-closing for more information). Also, please
write a comment on that issue linking back to this pull request once it is
open. -->

#### Brief description of what is fixed or changed
This PR implements the logical implications:

1) ```Q.gt(x, y)``` ⇒ ```Q.ge(x, y)```
2) ```Q.lt(x, y)``` ⇒ ```Q.le(x, y)```

These are currently returning ```None``` in ```ask()``` queries despite being trivially true. The fix introduces a direct implication check under the appropriate assumptions.

#### Implementation Details:
-A basic implication is checked using ```ask(Q.gt(lhs, rhs), assumptions)``` , and if it returns ```True```, we return ```True``` for the ```Q.ge``` (or ```Q.le```) query as expected.
-Unit tests have been added to confirm the correct behavior of these implications.
-Currently for some tests we get recursion error where **maximum recursion depth** exceeded error occurs for which reason this PR is marked as a draft.

#### Reasoning:
In SymPy, most relational predicates like ```Q.gt```, ```Q.lt```, ```Q.le``` are derived from ```Q.ge```(more specifically ```is_ge```). Therefore, we can usually reduce these queries to ```Q.ge```, but going the other way around (e.g., inferring Q.ge from Q.gt) is not handled directly. This PR fixes that gap using a basic logic check without touching any deeper structural assumptions logic.

#### Release Notes

<!-- Write the release notes for this release below between the BEGIN and END
statements. The basic format is a bulleted list with the name of the subpackage
and the release note for this PR. For example:

* solvers
  * Added a new solver for logarithmic equations.

* functions
  * Fixed a bug with log of integers. Formerly, `log(-x)` incorrectly gave `-log(x)`.

* physics.units
  * Corrected a semantical error in the conversion between volt and statvolt which
    reported the volt as being larger than the statvolt.

or if no release note(s) should be included use:

NO ENTRY

See https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more
information on how to write release notes. The bot will check your release
notes automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
* assumptions
  * ```ask(Q.ge(x),Q.gt(x))``` returns ```True``` instead of ```None```
  * ```ask(Q.le(x),Q.lt(x))``` returns ```True``` instead of ```None```

<!-- END RELEASE NOTES -->

krishnavbajoria02 committed:
Add check for explicitly checking for Q.gt predicate in assumptions

krishnavbajoria02 committed:
Added unit tests for inequality implications for Q.gt(),Q.ge(),Q.lt(),Q.le()

asmeurer commented:
I think we can just add a fact to the known facts `ge <=> ge & eq`.

krishnavbajoria02 commented:
Also I was looking into adding facts but I did realise that till now only unary predicate facts have been added so I will need to see more regarding how I can add facts relating to predicates dealing with more than one argument. Maybe that requires a new dictionary of its own ?

asmeurer commented:
I don't think it requires a separate dictionary. The underlying logic is still the same for this fact as with the unary facts. Namely, that the fact applies to any predicates that are applied over the same variables. So something like `Implies(Q.positive, Q.real)` means that whenever `Q.positive(x)` appears the predicate `Implies(Q.positive(x), Q.real(x))` can be added to the facts. The same thing applies here, only the facts have multiple arguments. This may already work, or you may just have to modify the code that processes known_facts to handle it.

krishnavbajoria02 committed:
Added realness_preds to capture realness of variables involved in Q.gt and Q.lt predicate and added unit tests

Added realness_preds to capture realness of variables involved in Q.gt and Q.lt predicate and added unit tests

krishnavbajoria02 committed:
Merge branch 'gt-ge-lt-le-fix' of https://github.com/krishnavbajoria02/sympy into gt-ge-lt-le-fix

krishnavbajoria02 committed:
Remove unnecessary checks in core/relational.py

krishnavbajoria02 commented:
> I don't think it requires a separate dictionary. The underlying logic is still the same for this fact as with the unary facts. Namely, that the fact applies to any predicates that are applied over the same variables. So something like `Implies(Q.positive, Q.real)` means that whenever `Q.positive(x)` appears the predicate `Implies(Q.positive(x), Q.real(x))` can be added to the facts. The same thing applies here, only the facts have multiple arguments. This may already work, or you may just have to modify the code that processes known_facts to handle it.

I tried adding this to the get all facts, but it currently doesn't support it so as you mentioned I will have to modify the code that processes known facts to handle it. 
Currently the above code does a good enough job for handling these cases. This is still not completely done but this should solve three issues :
1) ```Q.gt => Q.ge``` and ```Q.lt => Q.le```
2) ```Q.gt``` and ```Q.lt => Q.real``` for standalone variables, and 
3) ```Q.gt(x,y)``` and ```Q.gt(y,z) => Q.gt(x,z)```

krishnavbajoria02 committed:
Remove unused imports

krishnavbajoria02 committed:
Remove doctest failures

krishnavbajoria02 commented:
@asmeurer @TiloRC 
Please do review this PR. This does fix all the three above mentioned issues.
Thanks.

krishnavbajoria02 commented:
I also have tested out this code for working out a solution for queries of type ```ask(Q.zero(x),Q.eq(x,0))``` . It turns out this also could be handled by ```lra_satask``` if realness is implied so I have added another check for this.
Here is my approach :
```
def get_all_pred_and_expr_from_enc_cnf(enc_cnf):
    all_exprs = set()
    all_pred = set()
    realness_preds = set()
    for pred in enc_cnf.encoding.keys():
        if isinstance(pred, AppliedPredicate):
            if pred.function is Q.eq:
                lhs, rhs = pred.arguments
                if lhs.is_real and not rhs.is_real:
                    realness_preds.add(Q.real(rhs))
                if rhs.is_real and not lhs.is_real:
                    realness_preds.add(Q.real(lhs))
            if pred.function in (Q.gt, Q.lt):
                lhs, rhs = pred.arguments
                if lhs.is_Symbol:
                    realness_preds.add(Q.real(lhs))
                if rhs.is_Symbol:
                    realness_preds.add(Q.real(rhs))
            all_pred.add(pred)
            all_exprs.update(pred.arguments)

    return all_pred, all_exprs, realness_preds
```

asmeurer commented on sympy/assumptions/ask.py:
```diff
@@ -444,6 +444,7 @@ def ask(proposition, assumptions=True, context=global_assumptions):
     will not give a meaningful result.
 
     >>> ask(Q.positive(x), x > 0)
+    True
```
You should update the text of this note, not just the doctest.

krishnavbajoria02 replied:
Ok, I will do that.

krishnavbajoria02 replied:
I have updated the documentation. Please do review it. Thanks.

asmeurer commented:
I wonder if this can be done at a higher level than lra_satask. We want something like `Implies(x > y, Q.extended_real(x) & Q.extended_real(y))` to be added as a fact whenever `x > y` appears. @TiloRC would that implication be enough to make lra_satask work?

krishnavbajoria02 commented on sympy/assumptions/lra_satask.py:
```diff
@@ -216,14 +216,22 @@ def _pred_to_binrel(pred):
+                if lhs.is_Symbol:
+                    realness_preds.add(Q.real(lhs))
+                if rhs.is_Symbol:
+                    realness_preds.add(Q.real(rhs))
```
Yes but I think Tilo mentioned once that lra only works with real variable if I am not wrong like for positive and negative infinites there are some issues. Here is some more information in this comment : https://github.com/sympy/sympy/pull/28108.
But yes for correctness we can add extended real predicate in ```realness_preds```.

TiloRC replied:
It's incorrect to use real here, but I expect the system will raise exceptions if some of the variables are extended real. It might be fine to allow extended real variables, but it's something we should be careful about.

asmeurer commented:
I would also like adding an implication better than injecting assumptions. Injecting assumptions means you will get a confusing inconsistent assumptions error if one of the symbols isn't actually real. 

In general, I think we should work with the rule that an inequality is False if its arguments are nonreal. That should work better in the assumptions system, as long as we are careful to write facts that work with that. For instance, we cannot create a fact for symmetry like `(x >= y) | (y >= x)`. It has to be something like `Implies(Q.real(x) & Q.real(y), (x >= y) | (y >= x))`.

krishnavbajoria02 commented:
> I would also like adding an implication better than injecting assumptions. Injecting assumptions means you will get a confusing inconsistent assumptions error if one of the symbols isn't actually real.
> 
> In general, I think we should work with the rule that an inequality is False if its arguments are nonreal. That should work better in the assumptions system, as long as we are careful to write facts that work with that. For instance, we cannot create a fact for symmetry like `(x >= y) | (y >= x)`. It has to be something like `Implies(Q.real(x) & Q.real(y), (x >= y) | (y >= x))`.

Right. But I think if one of the variables are not real, ```lra_satask``` immediately returns ```UnhandledInput```. Isn't this behaviour desired?

krishnavbajoria02 commented:
> I would also like adding an implication better than injecting assumptions. Injecting assumptions means you will get a confusing inconsistent assumptions error if one of the symbols isn't actually real.

Also the inconsistent assumptions error won't occur here, due to not using ask when looking for Q.real(var) if var is not explicitly given as real.
As can be seen in this code:
```
if expr.is_real is not True:
            if expr.is_Symbol:
                if Q.real(expr) not in realness_preds: #does not use ask here, it just looks up in realness_preds set
                    raise UnhandledInput(f"LRASolver: {expr} must be real")
            else:
                raise UnhandledInput(f"LRASolver: {expr} must be real")
```

krishnavbajoria02 committed:
feat(assumptions): infer realness from equality; xfail multi-var inequalities; improve docs

krishnavbajoria02 commented:
> If I understand correctly, this only fixes the issue when the arguments of all the inequalities are a single symbol. Maybe it would be good to add some xfail tests for the broader cases of inequalities that don't work yet.

I have added those xfail tests. Please do review them. Thanks.

krishnavbajoria02 committed:
Added import for symbols for doctest

krishnavbajoria02 commented:
> I would also like adding an implication better than injecting assumptions. Injecting assumptions means you will get a confusing inconsistent assumptions error if one of the symbols isn't actually real.
> 
> In general, I think we should work with the rule that an inequality is False if its arguments are nonreal. That should work better in the assumptions system, as long as we are careful to write facts that work with that. For instance, we cannot create a fact for symmetry like `(x >= y) | (y >= x)`. It has to be something like `Implies(Q.real(x) & Q.real(y), (x >= y) | (y >= x))`.

@asmeurer 
I have followed the approach of adding facts in a separate PR (https://github.com/sympy/sympy/pull/28125) as it would get too clustered to have this approach in this PR only and also at least from my perspective this PR has some valuable additions so I wouldn't for the time being completely like to remove the changes made in this PR so far.

TiloRC commented:
> I wonder if this can be done at a higher level than lra_satask. We want something like Implies(x > y, Q.extended_real(x) & Q.extended_real(y)) to be added as a fact whenever x > y appears. @TiloRC would that implication be enough to make lra_satask work?

I don't think so. The theory solver treats `~(x > y)` as `(x <= y)`.

TiloRC commented on sympy/assumptions/tests/test_rel_queries.py:
```diff
@@ -159,6 +141,40 @@ def test_equality():
+    b = symbols('b', real=True)
+
+    # These should be inferable if simplification reduced the expression.
+    assert ask(Q.real(a), Q.gt(a + b, b + c)) is True
```
This doesn't look right. I'm pretty sure this assertion is incorrect.

TiloRC replied:
Consider a = (1+i), b = (1-i)

$$(1+i) + (1-i) = 2$$

and a = i, b = i

$$i * i = -1$$

asmeurer replied:
And note that this should be valid according to the `Implies(x > y, Q.extended_real(x) & Q.extended_real(y))` rule I suggested (https://github.com/sympy/sympy/pull/28108#issuecomment-2948107206). Except it also wouldn't work presently because:

```py
>>> print(ask(Q.real(a), Q.real(a + b) & Q.real(b)))
None
```

(we may need a better real theory solver to handle things like that better)

krishnavbajoria02 replied:
> Consider a = (1+i), b = (1-i)
> 
> ( 1 + i ) + ( 1 − i ) = 2
> 
> and a = i, b = i
> 
> i ∗ i = − 1

Actually in the inequality 
``` a+b>c+b ``` if b is assumed explicitly as real it would lead to cancelling of terms of b due to equality between them implying a>c which gives the fact that a and c are real. The above example shows why we should not add real predicates where lhs and rhs may contain more than a single variable.

TiloRC replied:
Didn't see real was assumed.

asmeurer commented on sympy/assumptions/lra_satask.py:
```diff
@@ -216,14 +216,28 @@ def _pred_to_binrel(pred):
         if isinstance(pred, AppliedPredicate):
+            if pred.function is Q.eq:
+                lhs, rhs = pred.arguments
+                if lhs.is_real and not rhs.is_real:
```
This `not rhs.is_real` will happen when it is None and when it is False. Is that what we want?

krishnavbajoria02 replied:
Right its something that might lead to raising inconsistent assumptions in case it is ```False```, but I think since we are using ```Q.eq``` so it should ideally raise an inconsistent assumptions (as lhs cant be equal to rhs where one side is real and the other is not) and for the case of None we want that to be added as a fact.

asmeurer replied:
It would be good to have tests for these different cases.

krishnavbajoria02 committed:
Refactor and integrate assumption extraction into check_satisfiability and added more unit tests

krishnavbajoria02 committed:
Remove unwanted functions

krishnavbajoria02 committed:
Added new line at the end of file

krishnavbajoria02 committed:
Fix error with additional new line

krishnavbajoria02 commented:
Are there any more changes needed to this PR ?

TiloRC commented on sympy/assumptions/lra_satask.py:
```diff
@@ -60,9 +117,47 @@ def check_satisfiability(prop, _prop, factbase):
+    ret = []
+    for expr in all_exprs:
+        if not hasattr(expr, "free_symbols"):
+            continue
+        if len(expr.free_symbols) == 0:
+            continue
+
+        if expr.is_real is not True:
+            if expr.is_Symbol:
+                if Q.real(expr) not in realness_preds:
+                    raise UnhandledInput(f"LRASolver: {expr} must be real")
+            else:
+                raise UnhandledInput(f"LRASolver: {expr} must be real")
+
+        # test for I times imaginary variable; such expressions are considered real
+        if isinstance(expr, Mul) and any(arg.is_real is not True for arg in expr.args):
+            raise UnhandledInput(f"LRASolver: {expr} must be real")
+
+        if expr.is_integer == True and expr.is_zero != True:
+            raise UnhandledInput(f"LRASolver: {expr} is an integer")
+        if expr.is_integer == False:
+            raise UnhandledInput(f"LRASolver: {expr} can't be an integer")
+        if expr.is_rational == False:
+            raise UnhandledInput(f"LRASolver: {expr} is irational")
+
+        if expr.is_zero:
+            ret.append(Q.zero(expr))
+        elif expr.is_positive:
+            ret.append(Q.positive(expr))
+        elif expr.is_negative:
+            ret.append(Q.negative(expr))
+        elif expr.is_nonzero:
+            ret.append(Q.nonzero(expr))
+        elif expr.is_nonpositive:
+            ret.append(Q.nonpositive(expr))
+        elif expr.is_nonnegative:
+            ret.append(Q.nonnegative(expr))
+
+    # Add new assumptions to both SAT encodings
+    for assm in ret:
```
I don't think it's a good idea to inline this function. I think this change makes the code much less readable. Also, this change seems unrelated to the rest of the PR.

krishnavbajoria02 replied:
Yeah,  I did this change because till I remember we had planned to get away with ```extract_pred_from_old_assum``` and ```get_all_pred_and_expr_from_enc_cnf``` as these were only being used once here. But if you want that we dont do this one inline and keep ```extract_pred_from_old_assum``` as it is I'll do that.

krishnavbajoria02 replied:
I have reverted back to the original code without trying to inline the above function. Please do review it. Thanks.

TiloRC committed:
Simplify realness deducing logic

TiloRC commented on sympy/assumptions/tests/test_rel_queries.py:
```diff
@@ -38,24 +38,6 @@ def test_lra_satask():
 
 
 def test_old_assumptions():
-    # test unhandled old assumptions
-    w = symbols("w")
-    raises(UnhandledInput, lambda: lra_satask(Q.lt(w, 2) & Q.gt(w, 3)))
-    w = symbols("w", rational=False, real=True)
-    raises(UnhandledInput, lambda: lra_satask(Q.lt(w, 2) & Q.gt(w, 3)))
-    w = symbols("w", odd=True, real=True)
-    raises(UnhandledInput, lambda: lra_satask(Q.lt(w, 2) & Q.gt(w, 3)))
-    w = symbols("w", even=True, real=True)
-    raises(UnhandledInput, lambda: lra_satask(Q.lt(w, 2) & Q.gt(w, 3)))
-    w = symbols("w", prime=True, real=True)
-    raises(UnhandledInput, lambda: lra_satask(Q.lt(w, 2) & Q.gt(w, 3)))
-    w = symbols("w", composite=True, real=True)
-    raises(UnhandledInput, lambda: lra_satask(Q.lt(w, 2) & Q.gt(w, 3)))
-    w = symbols("w", integer=True, real=True)
-    raises(UnhandledInput, lambda: lra_satask(Q.lt(w, 2) & Q.gt(w, 3)))
-    w = symbols("w", integer=False, real=True)
-    raises(UnhandledInput, lambda: lra_satask(Q.lt(w, 2) & Q.gt(w, 3)))
-
```
Why are these tests being removed?

TiloRC replied:
But all of these tests but the first one add have disallowed assumptions like prime=True

TiloRC commented on sympy/assumptions/lra_satask.py:
```diff
@@ -44,13 +44,70 @@ def lra_satask(proposition, assumptions=True, context=global_assumptions):
 
 def check_satisfiability(prop, _prop, factbase):
+    """
+    Determines the satisfiability of a given logical proposition `prop` and its negation `_prop`
+    under a provided fact base (`factbase`).
+
+    This function attempts to answer whether the proposition is always true, always false, or
+    indeterminate based on the current assumptions. It incorporates assumptions about realness,
+    rationality, and sign of the expressions.
```
I do like the Parameters / Returns / Raises sections though

krishnavbajoria02 committed:
Updated documentation and refactored functions

krishnavbajoria02 committed:
Remove doctest failure

TiloRC commented on sympy/assumptions/lra_satask.py:
```diff
@@ -258,7 +306,12 @@ def extract_pred_from_old_assum(all_exprs):
 
         if expr.is_real is not True:
-            raise UnhandledInput(f"LRASolver: {expr} must be real")
+            if expr.is_Symbol:
```
Why is important to check if expr is a Symbol?

TiloRC replied:
But if x + y aren't involved in query shouldn't that not matter?

krishnavbajoria02 committed:
Added back old assumptions tests

krishnavbajoria02 commented:
Please do let me know if any more changes are required in this PR. Thanks.

TiloRC commented on sympy/assumptions/lra_satask.py:
```diff
@@ -44,13 +44,75 @@ def lra_satask(proposition, assumptions=True, context=global_assumptions):
+        if not isinstance(pred, AppliedPredicate):
+            continue
 
+        if pred.function is Q.eq:
+            lhs, rhs = pred.arguments
+            for a, b in [(lhs, rhs), (rhs, lhs)]:
+                if a.is_real:
+                    if b.is_real is None:
+                        realness_preds.add(Q.real(b))
+                    elif b.is_real is False:
+                        raise ValueError("Inconsistent assumptions")
+        if pred.function in (Q.gt, Q.lt):
+            lhs, rhs = pred.arguments
+            if lhs.is_Symbol:
+                realness_preds.add(Q.real(lhs))
+            if rhs.is_Symbol:
+                realness_preds.add(Q.real(rhs))
```
Maybe add a signle comment for each of these heuristics. Maybe something like:
```python
# heuristic: x = y & real(x) -> real(y)
```

krishnavbajoria02 replied:
I have added these comments. Please do review them. Thanks.

krishnavbajoria02 committed:
Updated documentation

krishnavbajoria02 committed:
Removed unicode characters
